src/main/java/com/commercetools/sync/categories/helpers/CategoryBatchValidator.java

```diff
@@ -1,121 +1,98 @@
-            final List<CategoryDraft> categoryDrafts) {
-        Set<CategoryDraft> validDrafts = new HashSet<>();
-        Set<String> validKeys = new HashSet<>();
-        for (CategoryDraft categoryDraft : categoryDrafts) {
-            if (isValidCategoryDraft(categoryDraft)) {
-                ResourceIdentifier<Category> parent = categoryDraft.getParent();
-                if (parent == null) {
-                    validKeys.add(categoryDraft.getKey());
-                    validDrafts.add(categoryDraft);
-                } else if (isValidParentCategory(parent, categoryDraft.getKey())) {
-                    validDrafts.add(categoryDraft);
-                    validKeys.add(categoryDraft.getKey());
-                    validKeys.add(parent.getKey());
-                }
-            }
-        }
-
-        return ImmutablePair.of(validDrafts, validKeys);
+    @Override
+    public ImmutablePair<Set<CategoryDraft>, ReferencedKeys> validateAndCollectReferencedKeys(
+        @Nonnull final List<CategoryDraft> categoryDrafts) {
+        final ReferencedKeys referencedKeys = new ReferencedKeys();
+        final Set<CategoryDraft> validDrafts = categoryDrafts
+            .stream()
+            .filter(this::isValidCategoryDraft)
+            .peek(categoryDraft -> collectReferencedKeys(referencedKeys, categoryDraft))
+            .collect(Collectors.toSet());
+
+        return ImmutablePair.of(validDrafts, referencedKeys);
     }
 
-    private boolean isValidCategoryDraft(final CategoryDraft categoryDraft) {
-        String errorMessage = EMPTY;
-        if (categoryDraft != null) {
-            if (isBlank(categoryDraft.getKey())) {
-                errorMessage = format(CATEGORY_DRAFT_KEY_NOT_SET, categoryDraft.getName());
-            }
+    private boolean isValidCategoryDraft(@Nullable final CategoryDraft categoryDraft) {
+        if (categoryDraft == null) {
+            handleError(CATEGORY_DRAFT_IS_NULL);
+        } else if (isBlank(categoryDraft.getKey())) {
+            handleError(format(CATEGORY_DRAFT_KEY_NOT_SET, categoryDraft.getName()));
         } else {
-            errorMessage = CATEGORY_DRAFT_IS_NULL;
-        }
-
-        if (!isBlank(errorMessage)) {
-            handleError(errorMessage);
-            return false;
+            return true;
         }
 
-        return true;
+        return false;
     }
 
-    private boolean isValidParentCategory(final ResourceIdentifier<Category> parent, final String categoryDraftKey) {
-        if (isBlank(parent.getKey())) {
-            handleError(new SyncException(
-                    new ReferenceResolutionException(format(PARENT_CATEGORY_KEY_NOT_SET, categoryDraftKey))));
-            return false;
-        }
-
-        return true;
+    private void collectReferencedKeys(
+        @Nonnull final ReferencedKeys referencedKeys,
+        @Nonnull final CategoryDraft categoryDraft) {
+
+        referencedKeys.categoryKeys.add(categoryDraft.getKey());
```

Comment:
it's for all references to cache to avoid additional fetches, so here the category key could be a parent to another category so it's better to cache it too.

Reply:
@ahalberkamp you were right to question this! 

I will remove the category key in here which is unnecessary, as it is not needed for the reference resolution, categoryKeys should only collect parent references, similar to the other validator (for example product sync should only collect product keys in references). Second reason is it's unnecessary because the category key will be stored in cache but with other calls. (i.e fetchMatchingCategoriesByKeys)

Reply:
This was the commit: https://github.com/commercetools/commercetools-sync-java/pull/574/commits/c8ccd422a9a50e49adcc16c45881545841cd7868 but after that, I changed something more. adding here for reference.

Reply:
There is a special case for category sync, so I re-added draft key into categoryKeys, but this would be fixed when we use custom objects for the unresolved reference tracking like we have for transitions in state sync, and product reference attributes in product sync.
https://github.com/commercetools/commercetools-sync-java/pull/574/commits/31fdb392d571f6ea047c0f2fc2bb581e47237c21